### PR TITLE
i45: Onboarding Navigation

### LINKED ISSUE #45: Onboarding Navigation
Blocked by #44 

Once we have a navigation library up and running, we can use it for the onboarding flow. 

Here's a gif of the onboarding flow:

Sep-22-2020 12-26-05

**Requirements:**

- Add a checkpoint after user completes "Go to my location" step. If they quit the app and open it again, we should show the "Sign in with Twitter" button.
- If isSignedIn is true then don't show onboarding flow.
- ANDROID: The only onboarding screen with the ability to go back is the Twitter auth modal. For the rest, pressing the BACK button should close the app.

### PR DESCRIPTION
**Pull request by alfalcon90**
Fixes #45 

**TODO**

- [x] Call the `gotoLocation() `function in `components/Map.js` from `onboarding/index.js`
- [x] Fix Twitter functionality in `onboarding/index.js` then navigate to Home Screen
- [x] Check if user is logged in and skip onboarding screens

### CODE REVIEW

**Comment by alfalcon90** [comment]
This is what's currently implemented in onboarding/index.js right now.

Oct-18-2020 11-44-25

File: src/screens/onboarding/index.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    }
+  }
+
+  render () {
+    return (
+      <View style={styles.landscape}>
+        <Map></Map>
+        <View style={styles.safeArea} pointerEvents='box-none'>
+          <View style={styles.fabContainer}>
+            { !this.state.isLocated && <FAB label={R.strings.button.gotoLocation} theme='green' icon='crosshairs'
```
Comment (alfalcon90) [inline src/screens/onboarding/index.js:41]: TODO: Call the gotoLocation() function in components/Map.js from onboarding/location.js. 
If successful then -> this.setState({isLocated: true}) to switch FABs

File: src/screens/onboarding/index.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  render () {
+    return (
+      <View style={styles.landscape}>
+        <Map></Map>
+        <View style={styles.safeArea} pointerEvents='box-none'>
+          <View style={styles.fabContainer}>
+            { !this.state.isLocated && <FAB label={R.strings.button.gotoLocation} theme='green' icon='crosshairs' 
+              onPress={() => this.setState({isLocated: true})}/> }
```
Comment (alfalcon90) [inline src/screens/onboarding/index.js:42]: TODO: Call the gotoLocation() function in components/Map.js from onboarding/location.js. 
If successful then -> this.setState({isLocated: true}) to switch FABs

File: src/screens/onboarding/index.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            { this.state.isLocated && <FAB label={R.strings.button.logIn} theme='blue' icon='twitter' 
+              onPress={() => this.twitterLogin()
+                .then(() => {
+                  console.log('Signed in with Twitter!');
+                })
+                .catch(e => {
+                  console.log(e.message);
```
Comment (alfalcon90) [inline src/screens/onboarding/index.js:49]: Fix Twitter functionality then -> navigate to Home Screen

**Review by abdih — CHANGES_REQUESTED** [review]
I'm impressed with the level of contribution here; there is a lot of good stuff here.

I didn't get a chance to go through the entire PR because truthfully, I'm a bit mentally exhausted, but I do think I've raised some meaningful concerns.

Looking forward to working with you to fix some of them!

File: App.js [hunk 4]
```diff
@@ -4,30 +4,65 @@ import 'react-native-gesture-handler';
...
 
 const Stack = createStackNavigator();
 
-const App = () => {
+const App = (props) => {
+  // App launches
   useEffect(() => {
-    SplashScreen.hide();
+    // Load location from LS
+    getData('userLocation').then((location) => {
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply App.js:24]: Correct. isSignedIn gets updated (or not if the user is not signed in) by the call to the firebase library on line 31. Firebase handles their own state persistence as we discussed in our last chat so we don't store auth state in our own local storage.

File: src/library/components/TwitterAuth.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      // Perform the login request
+      const { authToken, authTokenSecret } = await RNTwitterSignIn.logIn();
+      // Shows activity indicator
+      props.onLoading(true)
+      // Create a Twitter credential with the tokens
+      const twitterCredential = auth.TwitterAuthProvider.credential(authToken, authTokenSecret);
+      // Sign-in the user with the credential
+      auth().signInWithCredential(twitterCredential);
+      // Update redux store
+      auth().onAuthStateChanged((user) => (user ? props.signIn(user) : null))
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/components/TwitterAuth.js:30]: This handles when the user first taps on the "Log in with Twitter" button. The one in App.js checks the auth state whenever the app launches.

File: App.js [hunk 6]
```diff
@@ -4,30 +4,65 @@ import 'react-native-gesture-handler';
...
+    // Load auth state
+    const subscriber = auth().onAuthStateChanged((user) => (user ? props.signIn(user) : null));
+
+    // App state listener
+    AppState.addEventListener('change', (e) => {
+      e !== 'unknown' ? props.updateAppState(e) : null
+    })
+
+    return () => {
+      subscriber; // unsubscribe on unmount
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply App.js:39]: I'm following the Firebase docs here: https://rnfirebase.io/auth/usage#listening-to-authentication-state
On closer inspection, what gets returned from calling `auth().onAuthStateChanged()` here is: `return () => subscription.remove();`

File: src/library/components/WorldMap.js [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+import { connect } from 'react-redux';
+import { setLoaded, updateUserVisible  } from 'rdx/actions';
+import useMapCamera from 'library/hooks/useMapCamera';
+import RES from 'res';
+
+const { MapView, Camera } = MapboxGL;
+MapboxGL.setAccessToken(config.MAPBOX_ACCESS_TOKEN);
+
+const WorldMap = (props) => {
+  const cameraRef = useRef(null)
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/components/WorldMap.js:17]: I thought we had an ESLint rule for it but somehow we never added one. Added rule and ran linter to fix all of these.

File: src/library/apis/geolocation.js [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+const config = {
+  enableHighAccuracy: true,
+  timeout: 15000,
+  maximumAge: 10000,
+  distanceFilter: 100,
+  interval: Platform.OS === 'android' ? 10000 : null,
+  fastestInterval: Platform.OS === 'android' ? 5000 : null,
+  useSignificantChanges: Platform.OS === 'ios' ? true : null,
+}
+let watchId = null;
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/apis/geolocation.js:76]: (1) We need the ID to be able to call `Geolocation.clearWatch(watchId)` in **stopWatchingLocation()**. It only gets used locally in this file so it doesn't really belong in the redux store, BUT it gets used in both **watchLocation()** and **stopWatchingLocation()** so that's why it's outside both functions.

File: src/library/apis/geolocation.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        user: [coords.longitude, coords.latitude],
+        hasPermission,
+      }
+      store.dispatch({ type: UPDATE_LOCATION, payload });
+      store.dispatch({ type: UPDATE_WATCHING, payload: true });
+    }
+
+    // Gate to prevent multiple instances
+    if (!watching) {
+      watchId = Geolocation.watchPosition(
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/apis/geolocation.js:97]: It took me a bit to remember why I did this in a strange way with two variables instead of one. The geolocation library starts with an ID of 0, which returns false when used in that if statement and you end up with two instances of **watchPosition()**.

File: src/library/apis/storage.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  try {
+    keys = await AsyncStorage.getAllKeys()
+  } catch (err) {
+    console.log(err)
+  }
+  try {
+    await AsyncStorage.multiRemove(keys)
+  } catch (err) {
+    console.log(err)
+  }
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/apis/storage.js:55]: I tried it at first but the description of it here scared me away: https://react-native-async-storage.github.io/async-storage/docs/api#clear

File: src/library/components/HomeBar.js [hunk 11]
```diff
@@ -58,6 +72,22 @@ const styles = create({
...
+    top: 78,
+    backgroundColor: RES.COLORS.asphaltGray,
+    borderWidth: 2,
+    borderColor: RES.COLORS.graniteGray,
+  },
 });
 
-export default HomeBar;
+const mapStateToProps = (state) => ({
+  photoURL: state.auth.user?.photoURL,
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/components/HomeBar.js:89]: We should be good for React Native, right?

File: src/library/components/WorldMap.js [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    mapLoadingHandler()
+
+    return () => {
+      isMounted = false;
+      props.setLoaded('map', false)
+    }
+  }, [])
+
+  // App is active || background
+  useEffect(() => {
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/components/WorldMap.js:47]: Found this: https://reactjs.org/docs/hooks-effect.html#tip-use-multiple-effects-to-separate-concerns

File: src/library/components/WorldMap.js [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+      <MapView
+        ref={mapRef}
+        animated={true}
+        style={styles.map}
+        styleURL={'mapbox://styles/alfalcon/cka1xbje712931ipd6i5uxam8'}
+        logoEnabled={false}
+        compassEnabled={false}
+        attributionEnabled={false}
+        onRegionIsChanging={(e) => regionChangeHandler(e)}
+        onDidFinishRenderingMapFully={!props.appState.loaded.map ? () => mapLoadingHandler(true) : null}
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/components/WorldMap.js:125]: It's confusing to me and I wrote it. I'll take a stab at refactoring.

File: App.js [hunk 14]
```diff
@@ -4,30 +4,65 @@ import 'react-native-gesture-handler';
...
+            <Stack.Screen name='Onboarding' component={OnboardingScreen} />
+          </>
+        )}
       </Stack.Navigator>
     </NavigationContainer>
   );
 };
 
-export default App;
+const mapStateToProps = (state) => ({ isSignedIn: state.auth.isSignedIn })
```
Comment (abdih) [inline App.js]: 🍎

File: App.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (abdih) [reply App.js:24]: the first part of what you're saying is true

> isSignedIn gets updated (or not if the user is not signed in) by the call to the firebase library on line 31

, but this part is not entirely accurate in representing the current implementation because as I'll elaborate in just a moment, (i) you are indeed storing auth state in our own local storage and (ii) you are redundantly telling firebase to do this storage.

> Firebase handles their own state persistence as we discussed in our last chat so we don't store auth state in our own local storage.

(i) on the line with 🍎 , you are saying that this component should be passed a property `isSignedIn` that is coming from our local storage: `state.auth.isSignedIn`

(ii) on the line with 🐱 , you are instructing how firebase should manipulate our local storage.

this flow to me is incorrect for at least 1 reason: if the user's login status changes, it will cause (ii) to be executed once more, since this component re-renders whenever mapStateToProps leads to a different result. In other words, you re-instruct how firebase should manipulate our local storage every time the user's login status changes, even though the instruction on _how_ firebase should manipulate the local storage doesn't change (i.e., firebase should still still listen to the auth state of the user and respectively set `state.auth.isSignedIn` to _true_ or _false_ based on that state).

Given the above, I see 1 change that I think we should act on now: the code for line (ii) should be moved 1 layer up to AppContainer, where the provider is specified.

The other change is one that I'm not sure is worth pursuing at this moment, but I'll express my thoughts on:

The react-redux library seems to essentially _force_ our hand to not be able to use `getData` as a means to construct the `props` for this component in `mapStateToProps` because this method is asynchronous in nature and according to the specifications of this library, this [isn't allowed in connect](https://react-redux.js.org/using-react-redux/connect-mapstate#mapstatetoprops-functions-should-be-pure-and-synchronous). This is not ideal to me because I'm not convinced there's any meaningful difference between the accessor for `isSignedIn` and `getData('userLocation')`, aside from the latter being a retrieval from an asynchronous variation of the local state (i.e., @react-native-async-storage/async-storage) and the former a synchronous one (i.e., rdx/store).

**Review by abdih — CHANGES_REQUESTED** [review]

File: src/library/components/WorldMap.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (abdih) [reply src/library/components/WorldMap.js:45]: There's a nuance here with the link you're referencing; `useEffect` does _not_ seem to specify any indication of when the effect should occur, where as in the link you referenced, the different methods have well-defined, segregated occurrences in the lifecycle of a component's rendering: https://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/

[the remark this answers is not in the document]
Comment (abdih) [reply src/library/components/WorldMap.js:45]: unless, the conditions inside these different effects essentially simulate the different canonical methods of a component, which I can see since you do use for instance `isMounted` in `App launches || quits`.

tl;dr: I'm convinced that this isn't a blocker and your comments above the `useEffects` are ok. I'm resolving this thread.

File: App.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (abdih) [reply App.js:24]: writing this as a request for pause on analyzing the above comment in the case I'm interpreting the line 🐱  incorrectly given your comment [here](https://github.com/Tinytown/tinytown-react-app/pull/53#discussion_r546272415).

File: App.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (abdih) [reply App.js:38]: the returned value that you've made explicit is not in contention with my original suggestion, since it's implying that `auth().onAuthStateChanged()` in fact returns an _unsubscriber_ -- this also implies that the variable is currently not named correctly. furthermore, the function that's returned in `useEffect`, according to the [docs](https://reactjs.org/docs/hooks-effect.html#example-using-hooks-1) sounds like is what will be used to clean up the component when it essentially unmounts. it's fine to technically unsubscribe listening to data about the user's login status once this component is unmounted, but that unsubscriber method should come in as a prop from AppContainer to also accommodate [this](https://github.com/Tinytown/tinytown-react-app/pull/53/files/851d5e428cbcbdb860a9e2adb71a8024788be156#r546434172) comment.

File: src/library/components/WorldMap.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/components/WorldMap.js]: Removed since it was no longer necessary.

File: src/library/components/TwitterAuth.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (abdih) [reply src/library/components/TwitterAuth.js]: I think there's some conflation here between subscription of the local state to firebase's data on the user's login status and the check against such a status. I think the subscription should happen in one place and at the AppContainer, with the relevant data flowing to components through mapStateToProps, and then inside various components, having the check against the respective attribute on that component's props.

please see [this](https://github.com/Tinytown/tinytown-react-app/pull/53/files/851d5e428cbcbdb860a9e2adb71a8024788be156#r546436931) comment for more context.

[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/components/TwitterAuth.js]: Ah this makes sense now. At some point, I think the instance in App.js was behind an if statement that prevented it from running all the time but yeah in this case we shouldn't need the second in TwitterAuth.js. I'll give it a shot. 

Should I move the location block right above it to AppContainer as well since it's basically doing the same thing but for user's location? I spent a lot of time trying to make this part of the redux/store.js code since it's basically setting the initial state of the store but there was no way out of async hell that I could figure out.

[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/components/TwitterAuth.js]: Fixed in a0b2a04

File: App.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (alfalcon90) [reply App.js]: The documentation that you get in VSCode for `onAuthStateChanged` has an example with your name recommendation and the actual method being called so maybe the example on the website is out-of-date. 
Screen Shot 2020-12-20 at 3 17 23 PM

File: src/library/components/WorldMap.js [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+  useEffect(() => {
+    let isMounted = true;
+    // Load static map and camera props
+    getMultiple(['cameraCenter', 'cameraZoom', 'userVisible']).then((res) => {
+      if (isMounted && res.cameraCenter) {
+        setCamera({
+          center: res.cameraCenter,
+          zoom: res.cameraZoom,
+          movedByUser: false });
+        props.updateUserVisible(res.userVisible);
```
Comment (abdih) [inline src/library/components/WorldMap.js:34]: usage of `updateUserVisibile` here and in a couple other places seems to disagree with the convention established in using a reducer in other parts of this change-set (e.g., store.dispatch(<function that'll return an action>(...)) or, for asynchronous local state interactions, using a method that has method signature `() => async (dispatch) => <returns action>`).

File: src/redux/actions/types.js [hunk 16]
```diff
@@ -0,0 +1,8 @@
+export const SIGN_IN = 'sign_in';
+export const SIGN_OUT = 'sign_out';
+export const UPDATE_LOCATION = 'update_location';
+export const UPDATE_WATCHING = 'update_watching';
+export const GO_TO_USER = 'go_to_user';
+export const USER_VISIBLE = 'user_visible';
+export const APP_STATE = 'app_state';
+export const APP_STORAGE = 'app_storage';
```
Comment (abdih) [inline src/redux/actions/types.js:8]: I think technically using [Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) is more rigorous here as it avoids accidental or mal-intentioned creation of an action.

Comment (abdih) [inline src/redux/actions/types.js:8]: I think technically using [Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) is more rigorous here as it avoids accidental or mal-intentioned creation of an action.

File: src/library/components/HomeBar.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (abdih) [reply src/library/components/HomeBar.js:89]: you're absolutely correct here; it looks like it's fully supported for react native: https://medium.com/bsadd/optional-chaining-in-react-native-805a374788d3

**Comment by abdih** [comment]
just pulled to checkout what the app looks like on the simulator for iOS, and it looks like (i) I'm unable to pinch, zoom, or pan, and (ii) when I rotate the device, it's as if the map rotates with the screen, which I don't believe is correct -- if I rotate the device so it's landscape right while the north side of the map is pointing to the north of SF, when the rotation is complete, I expect north to be in the 'up' direction still, even though the device's top is on the right.

File: src/library/apis/geolocation.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/apis/geolocation.js]: Moved to useLocation hook in f12ea60

File: src/library/apis/geolocation.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/apis/geolocation.js]: Moved to useLocation hook in f12ea60e9dc2e844d8f7c9d51f72c115c1268416

File: src/library/components/Scrim.js [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+    </View>
+  )
+}
+
+export default Scrim
+
+const styles = StyleSheet.create({
+  scrim: {
+    width: '100%',
+    marginTop: -40,
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/components/Scrim.js]: Fixed in d6b4a21b35dbb884efc52cf81704873c35f94dd7

**Comment by alfalcon90** [comment]
Pushed the last set of fixes and improvements, all set for reviews.

File: src/library/components/HomeBar.js [hunk 18]
```diff
@@ -1,31 +1,34 @@
...
+import { signOut } from 'rdx/authState';
+import { getLocation } from 'library/apis/geolocation';
 import { create } from 'library/utils/normalize.js';
 import { Menu, MenuDivider, MenuItem } from './Menu';
 import IconButton from './IconButton';
 import RES from 'res';
 
-const HomeBar = (props) => {
+const HomeBar = ({ signOut, goToUser, userVisible, photoURL }) => {
   const [showMenu, setShowMenu] = useState(false);
```
Comment (abdih) [inline src/library/components/HomeBar.js:15]: 🧨

File: src/screens/OnboardingScreen.js [hunk 19]
```diff
@@ -1,44 +1,44 @@
...
 import { create } from 'library/utils/normalize.js';
-import { MapView, TwitterAuth, FAB, ActivityOverlay } from 'library/components';
+import { WorldMap, TwitterAuth, FAB, ActivityOverlay } from 'library/components';
 import { withWait } from 'library/components/hoc';
 import RES from 'res';
 
-const OnboardingScreen = (props) => {
-  const [isLoading, setIsLoading] = useState(false);
+const OnboardingScreen = ({ storageLoaded, userVisible, goToUser }) => {
+  const [authLoading, setAuthLoading] = useState(false);
```
Comment (abdih) [inline src/screens/OnboardingScreen.js:12]: 🧨 
if it isn't there already and there's a good justification for this, could you please add to the CONTRIBUTING document what the convention is when we're supposed to use `useState` directly inside a component versus when we're supposed to use a more indirect approach (e.g., useMap) through _library/hooks_.

**Review by abdih — CHANGES_REQUESTED** [review]

File: src/library/hooks/useAppLaunch.js [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+import { updateAppState, getStateFromLS } from 'rdx/appState';
+import { signIn, updateAuth } from 'rdx/authState';
+
+export default (isSignedIn) => {
+  const [appIsReady, setAppIsReady] = useState(false);
+  const dispatch = useDispatch();
+
+  // Initial setup / start listeners
+  useEffect(() => {
+    dispatch(getStateFromLS());
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/hooks/useAppLaunch.js:15]: I'm a little confused, do you mean the first arrow function in `useEffect(() => {}` or the arrow function in the `getStateFromLS` declaration in appState.js?

File: src/library/components/WorldMap.js [hunk 21]
```diff
@@ -1,103 +1,30 @@
...
+import { create } from 'library/utils/normalize.js';
+import { updateUserVisible, updateUserLocation  } from 'rdx/locationState';
+import { useLocation, useMap } from 'library/hooks';
 import RES from 'res';
 
 const { MapView, Camera } = MapboxGL;
 MapboxGL.setAccessToken(config.MAPBOX_ACCESS_TOKEN);
 
-const WorldMap = (props) => {
+const WorldMap = ({ userLocation, updateUserLocation, updateUserVisible, children }) => {
```
[the remark this answers is not in the document]
Comment (alfalcon90) [reply src/library/components/WorldMap.js:15]: They're referencing the same function, they get passed as props down at the bottom by the connect function and then destructured here. https://react-redux.js.org/api/connect

File: src/screens/OnboardingScreen.js (hunk 19 above)
Comment (alfalcon90) [reply src/screens/OnboardingScreen.js:12]: Looking at the [React Hooks](https://reactjs.org/docs/hooks-custom.html#extracting-a-custom-hook) documentation it looks like the general rule of thumb is similar to any other function: "When we want to share logic between two JavaScript functions, we extract it to a third function". That's the case for useMap and useLocation. The other reason (which applies to useAppLaunch) is less about reusability and more about readability and keeping components focused on UI rendering while hooks handle the logic. I'll write up a version of this comment in the CONTRIBUTING file for future reference.
